source/pkger/pkger.go

```diff
@@ -0,0 +1,67 @@
+package pkger
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/golang-migrate/migrate/v4/source"
+	"github.com/golang-migrate/migrate/v4/source/httpfs"
+	"github.com/markbates/pkger/pkging"
+)
+
+func init() {
+	source.Register("pkger", &driver{})
+}
+
+// Instance is an implementation of http.FileSystem backed by an instance of
+// pkging.Pkger.
+type Instance struct {
+	pkging.Pkger
+
+	// Path is the relative path location of the migrations. It is passed to
+	// httpfs.PartialDriver.Init. If unset "/" is used as all paths are
+	// absolute.
+	Path string
+}
+
+// Open implements http.FileSystem.
+func (p *Instance) Open(name string) (http.File, error) {
+	f, err := p.Pkger.Open(name)
+	if err != nil {
+		return nil, err
+	}
+	return f.(http.File), nil
+}
+
+type driver struct {
```

Comment:
Why use a separate `driver` struct? You can embed `httpfs.PartialDriver` into `Pkger`

Reply:
Because the driver and the instance requires `Open` methods of different signatures.

* `source.Driver` requires `Open(url string) (source.Driver, error)`.
* `http.FileSystem` requires `Open(name string) (http.File, error)`.

It would've been nice if `pkging.Pkger` implemented  `http.FileSystem` but it turns out in fact it does not. I might be missing something but I need to implement `http.FileSystem` to be able to pass it to the partial driver `Init(fs http.FileSystem, path string) error` method.

`Pkger` (previously `Instance`) wraps `pkging.Pkger` and implements `http.FileSystem` so that it can be passed to the partial driver.

`driver` implements the `source.Driver` interface by providing its `Open` method.

If this is all unclear I'm happy to rework it somehow but you will have to provide some guidance on what you think makes more  sense.

I'm also  curious to know if I'm just missing something.

Reply:
You can use one struct by not embedding `pkging.Pkger`.

e.g.
```golang
type Pkger struct {
    httpfs.PartialDriver
    pkger pkging.Pkger
}
```

> It would've been nice if `pkging.Pkger` implemented `http.FileSystem` but it turns out in fact it does not.

Looks like [it](https://godoc.org/github.com/markbates/pkger/pkging#Pkger) [does](https://golang.org/pkg/net/http/#FileSystem)

> `Pkger` (previously `Instance`) wraps `pkging.Pkger` and implements `http.FileSystem` so that it can be passed to the partial driver.

You can pass `Pkger.pkger` to `Init()`

Reply:
> Looks like it does

The signatures are different. It does't return `http.File` but `pkging.File`. It's subtle but means
```
./pkger.go:56:28: cannot use instance.Pkger (type pkging.Pkger) as type http.FileSystem in argument to ds.PartialDriver.Init:
	pkging.Pkger does not implement http.FileSystem (wrong type for Open method)
		have Open(string) (pkging.File, error)
		want Open(string) (http.File, error)
```
What am I missing? I still don't see how it would work by simply not embedding `pkging.Pkger`.

If we want `Pkger` to be the driver and look like
```
type Pkger struct {
    httpfs.PartialDriver
    pkger pkging.Pkger
}
```
I'm not against that but I think we need to wrap `Pkger.pkger` to implement `http.FileSystem`. Another approach I can think of is to have
```
type fsFunc func(name string) (http.File, error)

func (f fsFunc) Open(name string) (http.File, error) {
  return f(name)
}
```
and then wrap `pkging.Pkger` just before calling `Init` like so
```
fs := fsFunc(func(name string) (http.File, error) {
	f, err := instance.pkger.Open(name)
	if err != nil {
		return nil, err
	}
	return f.(http.File), nil
})
if err := instace.Init(fs, instance.Path); err != nil {
```
What about that?

Reply:
Thinking about this a bit more I'm honestly wondering if it's worth adding this driver at all since it essentially only wraps `pkging.Pkger` in `http.FileSystem`. It might make sense to just put it on the user to wrap it themselves and instead use `httpfs.New` directly. What do you think about that?

For `packr.Box`, the other packager, it should be even easier since it alreay implements `http.FileSystem`, so the user can pass it straight in.

Reply:
> The signatures are different. It does't return `http.File` but `pkging.File` It's subtle but means ...

Oh wow, didn't notice that! That's annoying... Not sure why they went that route. At a *quick* glance, the interfaces look the same to me...
Looks like we'll need to wrap `pkging.Pkger`... I'm not picky about what `type Pkger struct` looks like as long as the fields aren't exported. 

> Thinking about this a bit more I'm honestly wondering if it's worth adding this driver at all since it essentially only wraps `pkging.Pkger` in `http.FileSystem`. It might make sense to just put it on the user to wrap it themselves and instead use `httpfs.New` directly. What do you think about that?

If you're only going to use `migrate` as a Go library, then `httpfs.New()` is probably the easiest and quickest route to get started. However, if you want to use DB URIs and the `migrate` CLI, you'll need to create a driver using the `httpfs.PartialDriver` and register it. You can't use `httpfs.New()` for this since doing so will hardcode the driver to the `http.FileSystem` specified via `httpfs.New()`.

I'd recommend having CLI support to debug and fix any issues you have with your migrations

Reply:
> However, if you want to use DB URIs and the migrate CLI, you'll need to create a driver using the httpfs.PartialDriver and register it. You can't use httpfs.New() for this since doing so will hardcode the driver to the http.FileSystem specified via httpfs.New().

I'm not sure I understand what you mean by hardcoding the driver to the specified file sytem. Could you elaborate?

How would the CLI support a source that is explicitly for embedded data? What would it mean to tell the CLI to read migrations from `pkger:///migrations`?

Looking at `go_bindata` it doesn't implement CLI support and as you can see I just did the  same by leaving `source.Driver.Open` unimplemented. Personally I only use `migrate` as a library.

That said, consumers of `pkger` mostly register resources in a global instance of `pkging.Pkger` using `pkger.Apply` and then access files using package scoped functions `pkger.Open`. This global instance it not, as far as I can tell, accessible. It seems like a good idea to use `source.Driver.Open` to return a driver that reads from the global `pkging.Pkger` instance. It would allow library users to opt for either `WithInstance` and their own instance of `pkging.Pkger` or for `Open` after registering migrations on the global instance with `pkger.Apply`.

I've actually already gone ahead ahead and implemented this along with increased test coverage. I'm happy to address any new feedback on that.

On the back of the above to partly answer my own question, `pkger:///migrations` will read migrations from the relative directory `/migrations` using the package scoped `pkger.Open` i.e. using a global instance of `pkging.Pkger`. However in the context of the CLI I'm not sure that helps anyone since they can't register migrations on the global instance that exists only in memory during the execution of the CLI, right?